# <img style="float: left; padding-right: 20px; width: 200px" src="https://raw.githubusercontent.com/raxlab/imt2200-data/main/media/logo.jpg">  IMT 2200 - Introducción a Ciencia de Datos
**Pontificia Universidad Católica de Chile**<br>
**Instituto de Ingeniería Matemática y Computacional**<br>
**Semestre 2025-S2**<br>
**Profesor:** Rodrigo A. Carrasco <br>

# <h1><center>Actividad 02: Extracción y Revisión de Datos</center></h1>

Esta actividad busca aplicar lo que acabamos de ver en clases sobre extracción de datos y conectarlo con la materia de "Tipos de Datos" que hemos visto en las clases pasadas.

## Instrucciones

Esto Notebook contiene las instrucciones a realizar para la actividad. 

<b>Al finalizarla, deben subir el Notebook y los archivos generados en un único archivo .zip, al módulo de la Actividad 02 en Canvas.</b>

Para esta actividad deben hacer lo siguiente:
<ol>
<li>Extraer datos</li>

Para esta actividad usaremos [**PokeAPI**](https://pokeapi.co/about), una interfaz de código abierto con información detallada relacionada a los videojuegos de Pokémon.
    
En esta actividad nos concentraremos en comparar las características básicas de los 151 pokémon originales. Para obtener la información de un pokémon, utilizaremos el siguiente URL:

```
https://pokeapi.co/api/v2/pokemon/{id_pokemon}
```

Donde `{id_pokemon}` es el identificador del pokémon (un número entero). 
    
<br>

<li>Formato</li>

¿En qué formato están los datos descargados? Identifique si son datos estructurados, semi-estructurados o no estructurados, explicando claramente por qué los categoriza así.

<br>
    
<li>DataFrame de análisis</li>

Construya un DataFrame con al menos las siguientes columnas: `name, type1, type2, hp, attack, defense, special-attack, special-defense, speed`. En caso de que un pokémon tenga solo un tipo, puede dejar su atributo `type2 = None`. Guarde este nuevo set de datos en un archivo CSV.

**Hint:** Inspeccione el atributo `stats` y `types` de las respuestas obtenidas. El valor que debe almacenar en el DataFrame estará bajó el nombre `base_stat`.
    
<br>
    
<li>Preguntas</li>

A partir de su nuevo dataset, responda las siguientes preguntas:

4.1 ¿Cuál son los 3 tipos (fuego, agua, tierra, etc.) más comunes en los pokémones de la primera generación?
    
4.2 ¿Cuál es el pokémon más rápido? ¿Y el más lento?

4.3 Genere una nueva columna en el DataFrame llamada `total-stats` que consista en la suma de todos los stats de un pokémon (`hp, attack, defense, special-attack, special-defense, speed`) y grafique un histograma de esta nueva variable. ¿Cuál es el rango de valores

## Rúbrica

- Si han hecho todo y sólo hay errores menores: 7.0
- Si sólo llegaron hasta la parte 3: 5.0
- Menos que eso: 1.0

## 1. Extraer datos

Utilizando el URL indicado en el enunciado, obtenga la información de los primeros 151 pokémones y guárdelos en una lista llamada `pokemones`. 

Para ello debe usar un `for` que itere sobre los diferentes IDs de pokemon y rescate la información de cada uno de ellos, concatenándola en una lista. Antes de iterar sobre los 151 pokemones, pruebe con 1 o 2 y una vez que verifique que su código está funcionando, aplique a los 151 primeros pokemones.

In [122]:
import requests


In [123]:
url_base = "https://pokeapi.co/api/v2/pokemon/" # URL base de la API
pokemones = [] # Inicializar lista para almacenar los datos de los pokémones

In [124]:
print("1. EXTRAYENDO DATOS DE POKÉMON...")

# Probarcon 2 pokémones para verificar que funciona la conexion con la API
print("Probando con 2 pokémones primero...")
for i in range(1, 3):
    try:
        url = f"{url_base}{i}"
        response = requests.get(url) # peticion GET a la URL con requests
        if response.status_code == 200:
            pokemon_data = response.json() # convertimos la respuesta a JSON para poder manejar los datos
            pokemones.append(pokemon_data) # agregamos los datos al final de la lista
            print(f"Pokémon {i} ({pokemon_data['name']}) extraído exitosamente")
        else:
            print(f"Error al extraer Pokémon {i}: {response.status_code}")
    except Exception as e:
        print(f"Error al procesar el pokémon {i}: {e}")
 
print(f"\nSe extrajeron {len(pokemones)} pokémones de prueba") # Total de pokemones extraidos para testear

1. EXTRAYENDO DATOS DE POKÉMON...
Probando con 2 pokémones primero...
Pokémon 1 (bulbasaur) extraído exitosamente
Pokémon 2 (ivysaur) extraído exitosamente

Se extrajeron 2 pokémones de prueba


In [125]:
# Ahora extraemos los 151 pokémones originales
print("\nExtrayendo los 151 pokémones...")
pokemones = []  # Reiniciamos la lista

for i in range(1, 152):
    try:
        url = f"{url_base}{i}"
        response = requests.get(url)
        if response.status_code == 200:
            pokemon_data = response.json()
            pokemones.append(pokemon_data)
            print(f"Pokémon {i} ({pokemon_data['name']}) extraído exitosamente")
        else:
            print(f"Error al extraer Pokémon {i}: {response.status_code}")
    except Exception as e:
        print(f"Error al procesar el pokémon {i}: {e}")

print(f"\nSe extrajeron {len(pokemones)} pokémones exitosamente") # imprimimos el total de pokemones extraidos
print(f"almacenados en la variable pokemones")


Extrayendo los 151 pokémones...
Pokémon 1 (bulbasaur) extraído exitosamente
Pokémon 2 (ivysaur) extraído exitosamente
Pokémon 3 (venusaur) extraído exitosamente
Pokémon 4 (charmander) extraído exitosamente
Pokémon 5 (charmeleon) extraído exitosamente
Pokémon 6 (charizard) extraído exitosamente
Pokémon 7 (squirtle) extraído exitosamente
Pokémon 8 (wartortle) extraído exitosamente
Pokémon 9 (blastoise) extraído exitosamente
Pokémon 10 (caterpie) extraído exitosamente
Pokémon 11 (metapod) extraído exitosamente
Pokémon 12 (butterfree) extraído exitosamente
Pokémon 13 (weedle) extraído exitosamente
Pokémon 14 (kakuna) extraído exitosamente
Pokémon 15 (beedrill) extraído exitosamente
Pokémon 16 (pidgey) extraído exitosamente
Pokémon 17 (pidgeotto) extraído exitosamente
Pokémon 18 (pidgeot) extraído exitosamente
Pokémon 19 (rattata) extraído exitosamente
Pokémon 20 (raticate) extraído exitosamente
Pokémon 21 (spearow) extraído exitosamente
Pokémon 22 (fearow) extraído exitosamente
Pokémon 23

In [126]:
# Verificar que se extrajeron los datos correctamente
print(f"\nTotal de pokémones extraídos: {len(pokemones)}") # Total de pokemones extraidos
print(f"Primer pokémon: {pokemones[0]['name']}") # Nombre del primer pokémon
print(f"\nEstructura del primer pokémon:") # Estructura del primer pokémon
print(f"Campos disponibles en un pokemon: {list(pokemones[0].keys())}") # Claves disponibles en los pokemones de la API

# Mostrar ejemplo de la estructura de un pokémon
print("\nEjemplo de la estructura de datos:") # Ejemplo de la estructura de datos
print(f"Nombre: {pokemones[0]['name']}") # Nombre del primer pokémon


for t in pokemones[0]['types']:
    print("\nTipos:") # Tipos del primer pokémon
    print(t['type']['name'])


for s in pokemones[0]['stats']:
    print("\nStats:") # Stats del primer pokémon
    print(s['stat']['name'], s['base_stat'])



Total de pokémones extraídos: 151
Primer pokémon: bulbasaur

Estructura del primer pokémon:
Campos disponibles en un pokemon: ['abilities', 'base_experience', 'cries', 'forms', 'game_indices', 'height', 'held_items', 'id', 'is_default', 'location_area_encounters', 'moves', 'name', 'order', 'past_abilities', 'past_types', 'species', 'sprites', 'stats', 'types', 'weight']

Ejemplo de la estructura de datos:
Nombre: bulbasaur

Tipos:
grass

Tipos:
poison

Stats:
hp 45

Stats:
attack 49

Stats:
defense 49

Stats:
special-attack 65

Stats:
special-defense 65

Stats:
speed 45


## 2. Formato de datos

Revise la lista que almacenó en `pokemones` y coméntenos en qué formato están los datos y qué tipos de datos serían estos (estructurados, semi-estructurados, no estructurados).

Respuesta:

In [127]:
print(f"Tipo: {type(pokemones[0])}") # tipo de datos de la variable pokemones
print(f"Es diccionario: {isinstance(pokemones[0], dict)}") # si es un diccionario
print(f"Es JSON válido: {isinstance(pokemones[0], (dict, list))}") # si es un JSON válido

Tipo: <class 'dict'>
Es diccionario: True
Es JSON válido: True


Los datos extraídos de la PokeAPI están en formato JSON, por lo tanto deberian ser considerdos como un tipo de dato semi estructurado, 
ya que JSON mantiene una estructura definida con campos específicos (name, types, stats), 
pero permite flexibilidad y anidación de datos como listas de objetos dentro de otros objetos,
características típicas de datos semi-estructurados que no siguen el formato rígido de tablas relacionales.

## 3. DataFrame de análisis

Construya ahor un DataFrame para poder analizar los datos posteriormente. Revise las columnas mínimas que debe tener según el enunciado inicial.

In [128]:
import pandas as pd

In [129]:
# Crear listas para almacenar los datos del DataFrame
names = []
type1s = []
type2s = []
hps = []
attacks = []
defenses = []
special_attacks = []
special_defenses = []
speeds = []

# Extraer datos de cada pokémon
for pokemon in pokemones:
    # Nombre
    names.append(pokemon['name'])
    
    # Tipos
    tipos = [t['type']['name'] for t in pokemon['types']]
    type1s.append(tipos[0] if len(tipos) > 0 else None)
    type2s.append(tipos[1] if len(tipos) > 1 else None)
    
    # Stats
    stats_dict = {s['stat']['name']: s['base_stat'] for s in pokemon['stats']}
    hps.append(stats_dict.get('hp', 0))
    attacks.append(stats_dict.get('attack', 0))
    defenses.append(stats_dict.get('defense', 0))
    special_attacks.append(stats_dict.get('special-attack', 0))
    special_defenses.append(stats_dict.get('special-defense', 0))
    speeds.append(stats_dict.get('speed', 0))

# Crear el DataFrame con los datos extraidos
df_pokemon = pd.DataFrame({
    'name': names,
    'type1': type1s,
    'type2': type2s,
    'hp': hps,
    'attack': attacks,
    'defense': defenses,
    'special-attack': special_attacks,
    'special-defense': special_defenses,
    'speed': speeds
})

print("DataFrame creado exitosamente:") 
print(f"Dimensiones: {df_pokemon.shape}")
print("\nPrimeras 5 filas:")
print(df_pokemon.head())

# Guardar el DataFrame en un archivo CSV
df_pokemon.to_csv('pokemon_stats.csv', index=False)
print("\nDataFrame guardado como 'pokemon_stats.csv'")

# Mostrar información del DataFrame
print("\nInformación del DataFrame:")
print(df_pokemon.info())

print("\nEstadísticas descriptivas:")
print(df_pokemon.describe())

DataFrame creado exitosamente:
Dimensiones: (151, 9)

Primeras 5 filas:
         name  type1   type2  hp  attack  defense  special-attack  \
0   bulbasaur  grass  poison  45      49       49              65   
1     ivysaur  grass  poison  60      62       63              80   
2    venusaur  grass  poison  80      82       83             100   
3  charmander   fire    None  39      52       43              60   
4  charmeleon   fire    None  58      64       58              80   

   special-defense  speed  
0               65     45  
1               80     60  
2              100     80  
3               50     65  
4               65     80  

DataFrame guardado como 'pokemon_stats.csv'

Información del DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 151 entries, 0 to 150
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   name             151 non-null    object
 1   type1            151 non-nul

## 4. Preguntas

#### 4.1 ¿Cuál son los 3 tipos (*water, fire, ground,* etc.) más comunes en los pokémones de la primera generación?

In [130]:
# 4.1 Tipos más comunes

# Contar tipos 
tipos_combinados = df_pokemon['type1'].tolist() + df_pokemon['type2'].dropna().tolist()
print(tipos_combinados) # tipos combinados
conteo_tipos = pd.Series(tipos_combinados).value_counts()
print(conteo_tipos) # conteo de tipos

['grass', 'grass', 'grass', 'fire', 'fire', 'fire', 'water', 'water', 'water', 'bug', 'bug', 'bug', 'bug', 'bug', 'bug', 'normal', 'normal', 'normal', 'normal', 'normal', 'normal', 'normal', 'poison', 'poison', 'electric', 'electric', 'ground', 'ground', 'poison', 'poison', 'poison', 'poison', 'poison', 'poison', 'fairy', 'fairy', 'fire', 'fire', 'normal', 'normal', 'poison', 'poison', 'grass', 'grass', 'grass', 'bug', 'bug', 'bug', 'bug', 'ground', 'ground', 'normal', 'normal', 'water', 'water', 'fighting', 'fighting', 'fire', 'fire', 'water', 'water', 'water', 'psychic', 'psychic', 'psychic', 'fighting', 'fighting', 'fighting', 'grass', 'grass', 'grass', 'water', 'water', 'rock', 'rock', 'rock', 'fire', 'fire', 'water', 'water', 'electric', 'electric', 'normal', 'normal', 'normal', 'water', 'water', 'poison', 'poison', 'water', 'water', 'ghost', 'ghost', 'ghost', 'rock', 'psychic', 'psychic', 'water', 'water', 'electric', 'electric', 'grass', 'grass', 'ground', 'ground', 'fighting', 

In [131]:
print("Los 3 tipos más comunes en la primera generación son:")
print(conteo_tipos.head(3))

# También mostrar solo type1 para comparar
print("\nTipos más comunes solo considerando type1:")
print(df_pokemon['type1'].value_counts().head(3))



Los 3 tipos más comunes en la primera generación son:
poison    33
water     32
normal    22
Name: count, dtype: int64

Tipos más comunes solo considerando type1:
type1
water     28
normal    22
poison    14
Name: count, dtype: int64


El tipo más común es: WATER (agua)

#### 4.2 ¿Cuál es el pokémon más rápido? ¿Y el más lento?

In [132]:
# 4.2 Pokémon más rápido y más lento

pokemon_speed = df_pokemon['speed'] # velocidad de los pokemones
print(f"La velocidad de los pokemones es: {pokemon_speed}")

pokemon_speed_max = pokemon_speed.max() # velocidad maxima de los pokemones
print(f"La velocidad máxima de los pokemones es: {pokemon_speed_max}")

pokemon_speed_min = pokemon_speed.min() # velocidad minima de los pokemones
print(f"La velocidad mínima de los pokemones es: {pokemon_speed_min}")

La velocidad de los pokemones es: 0       45
1       60
2       80
3       65
4       80
      ... 
146     50
147     70
148     80
149    130
150    100
Name: speed, Length: 151, dtype: int64
La velocidad máxima de los pokemones es: 150
La velocidad mínima de los pokemones es: 15


In [133]:
# mostrar los 10 pokemones mas rapidos
print(df_pokemon.nlargest(10, 'speed').head())
# mostrar los 10 pokemones mas lentos
print(df_pokemon.nsmallest(10, 'speed'))

           name     type1   type2   hp  attack  defense  special-attack  \
100   electrode  electric    None   60      50       70              80   
134     jolteon  electric    None   65      65       60             110   
141  aerodactyl      rock  flying   80     105       65              60   
149      mewtwo   psychic    None  106     110       90             154   
50      dugtrio    ground    None   35     100       50              50   

     special-defense  speed  
100               80    150  
134               95    130  
141               75    130  
149               90    130  
50                70    120  
           name   type1    type2   hp  attack  defense  special-attack  \
78     slowpoke   water  psychic   90      65       65              40   
38   jigglypuff  normal    fairy  115      45       20              45   
73      geodude    rock   ground   40      80      100              30   
45        paras     bug    grass   35      70       55              45   

In [134]:
# mostrar el pokemon mas rapido
print(df_pokemon.loc[df_pokemon['speed'].idxmax()])

# mostrar el pokemon mas lento
print(df_pokemon.loc[df_pokemon['speed'].idxmin()])

name               electrode
type1               electric
type2                   None
hp                        60
attack                    50
defense                   70
special-attack            80
special-defense           80
speed                    150
Name: 100, dtype: object
name               slowpoke
type1                 water
type2               psychic
hp                       90
attack                   65
defense                  65
special-attack           40
special-defense          40
speed                    15
Name: 78, dtype: object


In [135]:
print(f"El pokemon mas rapido es: {df_pokemon.loc[df_pokemon['speed'].idxmax()]['name']}")
print(f"La velocidad de {df_pokemon.loc[df_pokemon['speed'].idxmax()]['name']} es: {df_pokemon.loc[df_pokemon['speed'].idxmax()]['speed']}")


print(f"El pokemon mas lento es: {df_pokemon.loc[df_pokemon['speed'].idxmin()]['name']}")
print(f"La velocidad de {df_pokemon.loc[df_pokemon['speed'].idxmin()]['name']} es: {df_pokemon.loc[df_pokemon['speed'].idxmin()]['speed']}")

El pokemon mas rapido es: electrode
La velocidad de electrode es: 150
El pokemon mas lento es: slowpoke
La velocidad de slowpoke es: 15


El pokémon más rápido es: Electrode

#### 4.3 Genere una nueva columna en el DataFrame llamada `total-stats` que consista en la suma de todos los stats de un pokémon (`hp, attack, defense, special-attack, special-defense, speed`) ¿Cuáles son los valores mínimo y máximo de esta nueva variable?

In [136]:
# 4.3 Columna total-stats
# Crear la columna total-stats
df_pokemon['total-stats'] = (df_pokemon['hp'] + df_pokemon['attack'] + 
                             df_pokemon['defense'] + df_pokemon['special-attack'] + 
                             df_pokemon['special-defense'] + df_pokemon['speed'])

print("Columna 'total-stats' creada exitosamente")
print(f"\nRango de valores:")
print(f"Valor mínimo: {df_pokemon['total-stats'].min()}")
print(f"Valor máximo: {df_pokemon['total-stats'].max()}")
print(f"Valor promedio: {df_pokemon['total-stats'].mean():.2f}")

# Mostrar los pokémones con stats totales más altos y más bajos
print("\nPokémon con stats totales más altos:")
top_stats = df_pokemon.nlargest(5, 'total-stats')[['name', 'type1', 'type2', 'total-stats']]
print(top_stats)

print("\nPokémon con stats totales más bajos:")
bottom_stats = df_pokemon.nsmallest(5, 'total-stats')[['name', 'type1', 'type2', 'total-stats']]
print(bottom_stats)

Columna 'total-stats' creada exitosamente

Rango de valores:
Valor mínimo: 195
Valor máximo: 680
Valor promedio: 407.64

Pokémon con stats totales más altos:
          name     type1   type2  total-stats
149     mewtwo   psychic    None          680
148  dragonite    dragon  flying          600
150        mew   psychic    None          600
143   articuno       ice  flying          580
144     zapdos  electric  flying          580

Pokémon con stats totales más bajos:
         name  type1   type2  total-stats
9    caterpie    bug    None          195
12     weedle    bug  poison          195
128  magikarp  water    None          200
10    metapod    bug    None          205
13     kakuna    bug  poison          205


El valor máximo lo tiene: Mewtwo (total-stats: 680)
y el mínimo lo tiene: Caterpie (total-stats: 195)

#### 4.4 Grafique un histograma de `total-stats` y comente brevemente lo que observa de la distribución de este atributo.

In [ ]:
import matplotlib.pyplot as plt





ModuleNotFoundError: No module named 'matplotlib'

Descripción: